src/DonationForms/resources/app/utilities/usePubSub.ts

```diff
@@ -0,0 +1,66 @@
+import {RefObject, useEffect} from 'react';
+
+/**
+ * @unreleased
+ */
+export default function usePubSub() {
+
+    const events = {};
+
+    useEffect(() => {
+        addEventListener('message', eventListener, false);
+        return () => removeEventListener('message', eventListener);
+    }, []);
+
+    const getMessage = (message: MessageEvent) => {
+        // handle IframeResizer messages
+        if (typeof message.data === 'string') {
+            if (message.data.includes('[iFrameSizer]message:')) {
+                return JSON.parse(message.data.replace('[iFrameSizer]message:', ''));
+            }
+        }
+
+        return message.data
+    }
+
+    const eventListener = (message: MessageEvent) => {
+        const {event, data} = getMessage(message);
+
+        if (events[event]) {
+            events[event].forEach((callback: (data: any) => void) => {
+                callback(data);
+            });
+        }
+    }
+
+    const subscribe = (event: string, callback: Function) => {
+        if (!events[event]) {
+            events[event] = [];
+        }
+
+        events[event].push(callback);
+    }
+
+    const publish = (event: string, data: any, iframeRef?: RefObject<any>) => {
+        const message = {
+            event,
+            data
+        }
+
+        if (iframeRef) {
+            // IframeResizer
+            if (iframeRef.current?.sendMessage) {
+                iframeRef.current.sendMessage(message);
+            } else {
+                iframeRef.current?.contentWindow?.postMessage(message);
+            }
+        } else {
+            postMessage(message);
+        }
+    }
+
+    return {
+        publish,
+        subscribe
+    }
```

Comment:
Let's add an `unsubscribe` method as well so we can handle cleanup for events.

Reply:
It is not a problem to add, but I don't think it's necessary since we are using this only in the context of the donation form preview. Even if we add that,  how it will work? I mean, right now we don't have a global state of events, so this will work only in the context of the component where we are subscribed to an event.

Reply:
It's simply for handling re-rendering in React inside of `useEffect`. We're relying on `preview`, but not including it as a dependency, and if the component re-renders we'd be accumulating subscriptions. We want to return a function in the `useEffect` which handles the downward trigger.